### PR TITLE
Detect aliasing in cuDNN RNN flatten_parameters

### LINKED ISSUE #3751: cuDNN RNNs silently break weight tying in __init__
While the fast path for cuDNN RNNs introduced in https://github.com/pytorch/pytorch/pull/2179 includes a check intended to fall back to copying weights if weights are moved or aliased, this only works if the aliasing happens after the weights are first compacted. But the most natural place to tie weights is inside a module's `__init__` method, which runs before `.cuda()`, the typical time when `flatten_parameters` is first invoked.

Here's an MWE. The three runs (CPU, weight tying before `.cuda()`, and weight tying after `.cuda()`) should all give the same results, and they do before #2179, but they have not since:

```python
import torch

for cuda in (False, True):
    print('CUDA:', cuda)
    for when in ('before', 'after') if cuda else ('before',):
        if cuda: print('tying', when, '.cuda()')
        torch.manual_seed(1)
        rnn = torch.nn.LSTM(1, 1, num_layers=1, bidirectional=True, dropout=0)
        if when == 'before': rnn.bias_ih_l0_reverse = rnn.bias_ih_l0
        inputs = torch.autograd.Variable(torch.rand(1, 1, 1))
        if cuda:
            inputs = inputs.cuda(0)
            rnn.cuda(0)
        if when == 'after': rnn.bias_ih_l0_reverse = rnn.bias_ih_l0
        opt = torch.optim.SGD(rnn.parameters(), lr=0.1)

        opt.zero_grad()
        output = rnn(inputs)[0]
        print('iter 1 output:', output.data[0, 0, 0])
        output.sum().backward()

        opt.step()
        output = rnn(inputs)[0]
        print('iter 2 output:', output.data[0, 0, 0])
```

### PR DESCRIPTION
**Pull request by jekbradbury**
Fixes #3751. 
Currently, `nn.RNN.flatten_parameters` sets up a contiguous weight buffer to enable the no-copy code path for cuDNN RNNs, then uses `set_` to redirect the data pointers within the RNN's parameters to point into that weight buffer. Lastly it caches the list of parameter pointers within the buffer so that `nn.RNN.forward` can detect if the RNN's parameters change and fall back to the slow path. If certain preconditions aren't met, the cached list of parameter pointers is empty, forcing the RNN to always use the slow path. This patch adds an aliasing check to those preconditions.

The added test fails on master and succeeds with this change.

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -66,6 +66,15 @@ def flatten_parameters(self):
             self._data_ptrs = []
             return
 
+        # If any parameters alias, we fall back to the slower, copying code path. This is
+        # a sufficient check, because overlapping parameter buffers that don't completely
+        # alias would break the assumptions of the uniqueness check in
+        # Module.named_parameters().
+        data_ptrs = list(p.data.data_ptr() for layer in self.all_weights for p in layer)
+        if len(set(data_ptrs)) != len(data_ptrs):
```
Comment (apaszke) [inline torch/nn/modules/rnn.py]: It would be faster to construct it as a `set` and then compare to length of `self.all_weights`

**Review by apaszke — COMMENTED** [review]
Looks good! One minor comment

File: torch/nn/modules/rnn.py (hunk 1 above)
Comment (jekbradbury) [reply torch/nn/modules/rnn.py]: `self.all_weights` is nested, but I just moved the non-nested `all_weights` up from later in the code, so this should still be faster overall

Comment (jekbradbury) [reply torch/nn/modules/rnn.py]: self.all_weights is nested, but I just changed to something that may be slightly faster?

**Comment by soumith** [comment]
thanks James!

**Comment by Johnqczhang** [comment]
After I update pytorch to the latest version and recompile, I met the following warning when I run an LSTM model:
"UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greately increasing memory usage. To compact weights again call flatten_parameters()."

**Comment by apaszke** [comment]
Yup, this is caused by a bug from another PR, which makes these checks always fail. It's fixed in #3853 .

**Pull request merged**